File: exercism/clojure/robot-name/src/robot_name.clj
(ns robot-name)

(def used-names (ref #{}))
(defn- rand-uppercase-letter []
  (-> (inc (- 90 65))
      (rand-int)
      (+ 65)
      (char)))

(defn- random-name-simple []
  (format "%s%s%d" (rand-uppercase-letter) (rand-uppercase-letter) (+ 100 (rand-int 900))))


(defn- random-name []
  (let [new-name (ref (random-name-simple) ) ]
    (while (contains? @used-names @new-name)
      (ref-set new-name (random-name-simple))
      )
    (dosync (alter used-names conj @new-name))
    @new-name
    )
  )


(defn robot []
  (ref (random-name))
)

(defn robot-name [robot]
      @robot
)

(defn reset-name [robot] ;; <- arglist goes here
  (dosync (ref-set robot (random-name)))
)

File: exercism/julia/robot-name/robot-name.jl
const usedNames = Set{String}()
mutable struct Robot    
    name::String    
end

function Robot()
    Robot(makeuniquename())
end

function makeuniquename()::String
    new_name = cat(rand('A':'Z',2), rand('0':'9',3), dims = 1) |> join
    while new_name ∈ usedNames
        new_name = cat(rand('A':'Z',2), rand('0':'9',3), dims = 1) |> join
    end
    push!(usedNames, new_name)
    new_name
end

function reset!(instance::Robot)
    instance.name = makeuniquename()
end

function name(instance::Robot)
    instance.name
end
